Simplest version of console logging each and every interaction

mmocny opened the pull request.

tunetheweb commented:
I think it's confusing to expose an interaction as "INP" when it's not the INP. For example:

<img width="257" alt="image" src="https://user-images.githubusercontent.com/10931297/235227421-3cbd8061-3865-49fd-8c71-619d900be180.png">

The second one is less than the first one so is clearly not a new INP, but we've labelled it as INP.

So I think we should continue to log the INP metric, but log this as a separate "interaction" metric. Either always like this:

```
[Web Vitals Extension] INP 48: Interaction: 48
[Web Vitals Extension] INP 48: Interaction: 40
```

Or on separate lines like this
```
[Web Vitals Extension] Interaction: 48
[Web Vitals Extension] INP 48
[Web Vitals Extension] Interaction: 40
```

Note the second one has duplicated the first one (once as an "Interaction" and once as the "INP" value. Not sure if we can combine these:
```
[Web Vitals Extension] INP Interaction: 48
[Web Vitals Extension] Interaction: 40
```

WDYT?

I'm in two minds about exposing this as a separate config setting. I like the simplicity of not doing that, but worry about how noisy it gets on a highly interactive site. Will think on it some more...

mmocny commented:
- Updated the strings to be "Interaction" not "INP"
- Changed onEachInteraction to group interactions by interactionID, to reduce logging.
  - web-vitals.js already does this for INP.
- Because of that, I also changed to console.table a breakdown table *for each part*
  - also do this for INP.  I think this is an improvement, see below:

![Screenshot 2023-04-28 at 4 47 58 PM](https://user-images.githubusercontent.com/474282/235251510-86d58e3d-54fa-4cea-8e76-6b0b9ccc90c1.png)


One more feature to go: reduce the duplication when you get a new INP.

mmocny commented:
Done.  Found a surprisingly simple solution.

brendankenny commented:
- fixed some minor lint stuff (092a60f)
- tracked down the negative durations (a0c1664)

brendankenny commented on src/browser_action/onEachInteraction.js:
```diff
@@ -0,0 +1,42 @@
+export function onEachInteraction(callback) {
+	const valueToRating = score => score <= 200 ? 'good' : score <= 500 ? 'needs-improvement' : 'poor';
```
It's possible but I don't think worth jumping through the extra hoops right now. We should maybe consider this mode helpful enough to upstream something like in into web-vitals.js (`onInteraction`?)

tunetheweb replied:
Let’s wait and see how this lands and how @philipwalton feels. Know he’s been keen on the past on leaving this as a CWV library rather than something more general.

Plus LoAF might change things and adding this to the library would make any change with that a breaking change.

mmocny replied:
Ah, fair, the current logging (especially breakdowns) would have been easier to do elegantly with LoAF.  It may be that that implementation is also a lot less bytes to add to the library.

brendankenny commented on src/browser_action/on-each-interaction.js:
```diff
@@ -0,0 +1,58 @@
+
+  observer.observe({
+    type: 'event',
+    durationThreshold: 0,
```
One mismatch is that this is using `durationThreshold: 0` but the INP logger is using the `durationThreshold: 40` default. It's not necessarily a problem, but it does mean the `Interaction` breakdowns tend to have quite a few more subparts than the `INP` breakdowns.

mmocny replied:
I think the 40 thing we should consider changing down to 0.
40 was deemed to picked just to try and limit the perf cost of the extension but I haven't seen any real measure effect-- but I did see examples of folks being confused that INP doesn't report after interaction.

tunetheweb replied:
SGTM. Though think we should keep web-vitals at 40 as likely will be a bigger perf concern there.

tunetheweb commented on src/browser_action/on-each-interaction.js:
```diff
@@ -0,0 +1,58 @@
+      const entry = interaction.reduce((prev, curr) => prev.duration >= curr.duration ? prev : curr);
+      const value = entry.duration;
+
+      callback({
+        attribution: {
+          eventEntry: entry,
+          eventTarget: '',
+          eventTime: entry.startTime,
+          eventType: entry.name,
+          loadState: 'unknown',
+        },
+        delta: 0,
+        entries: interaction,
+        id: 'none',
+        name: 'Interaction',
+        navigationType: 'unknown',
+        rating: valueToRating(value),
+        value,
```
Let's some of these that won't be set. I don't think it needs to be exactly the same as INP entry and worry about people expecting these to be set.

```suggestion
      callback({
        attribution: {
          eventEntry: entry,
          eventTime: entry.startTime,
          eventType: entry.name,
        },
        entries: interaction,
        name: 'Interaction',
        rating: valueToRating(value),
        value,
```

tunetheweb reviewed (approved):
OK LGTM with two further suggestions.

WDYT @brendankenny / @mmocny ?

mmocny commented:
I accepted the suggestion to remove the INP Interaction comparison, because I won't have the time to make improvements.

- onINP is registered before onEachInteraction, and so the perf observer callback should come first
- However, if there is any async "hop" for dispatching the callback, onEachInteraction will fire its callback first.
  - For example, in a prior iteration of this patch we would async hop for `chrome.storage` and I had to make sure not to do that.
  - `onINP` in web-vitals.js seems to unconditionally call `report()` synchronously, but I'm not sure if `bindReporter` has some reasons to delay dispatch...

tunetheweb commented:
Cool. Good to merge from your side @mmocny ?

Will still wait until @brendankenny gives his OK too in case he has any last comments.

mmocny commented:
Yep! Thanks!

On Tue, May 2, 2023, 13:59 Barry Pollard ***@***.***> wrote:

> Cool. Good to merge from your side @mmocny <https://github.com/mmocny> ?
>
> Will still wait until @brendankenny <https://github.com/brendankenny>
> gives his OK too in case he has any last comments.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleChrome/web-vitals-extension/pull/120#issuecomment-1531345540>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADTZKVI5EPE2UZTZTMAIYDXEDZLRANCNFSM6AAAAAAXPQXN3A>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

brendankenny reviewed (approved):
LGTM!
